### DISCUSSION TITLE
Why use XSync in X11FramebufferSurface.Blit

### DISCUSSION THREAD

**Discussion by lindexi** [AvaloniaUI/Avalonia#16690] [Q&A]
As the code shows:

https://github.com/AvaloniaUI/Avalonia/blob/37ba4ce58e2e9d251b21310019fda40b81099b19/src/Avalonia.X11/X11FramebufferSurface.cs#L25-L52

The `X11FramebufferSurface.Blit` will use XSync not the XFlush here. And as I know, the XSync will sync the X11 client-server input and output, which will latency here. And the `X11FramebufferSurface.Blit` do is put the Image from XClient to XServer, and I think the action will be Flush.

Any help or guidance would be greatly appreciated. Thank you.

**Comment by maxkatz6** [comment]
cc @kekekeks

**Comment by kekekeks** [comment]
Looks like a leftover code. Since now we are using a separate X11 connection for presentation, XFlush should be enough.

**Reply by kekekeks** [reply]
It might be required for backpressure though, so we might want to have an XSync call after we finish processing the composition commit for all windows.

**Reply by kekekeks** [reply]
i. e. on low power devices we might be sending images faster than XServer processes them.

**Reply by kekekeks** [reply]
We don't need that for OpenGL rendering since `glxMakeCurrent` provides such backpressure automatically when we pass it a XID that has an unfinished `glxSwapBuffers` call

**Reply by kekekeks** [reply]
Another point about XSync/XFlush is potential pending requests arriving after the window is closed by the UI thread. So we still need to issue an XSync when disposing X11FramebufferSurface

**Reply by lindexi** [reply]
@kekekeks Thank you. The XLib will copy the image data in XPutImage which means the X11FramebufferSurface disposing will be safe.

**Reply by lindexi** [reply]
What puzzles me is that, as https://stackoverflow.com/a/31090588 says, Avalonia not need call XSync here.

> Finally, if you passed False, XSync does not discard the events in the queue. If you passed True, XSync discards all events in the queue, including those events that were on the queue before XSync was called. Client applications seldom need to call XSync.

**Reply by kekekeks** [reply]
>The XLib will copy the image data in XPutImage which means the X11FramebufferSurface disposing will be safe.

1) Window gets closed on the UI thread
2) UI thread waits for the last pending render operation to complete
3) X11FramebufferSurface.Blit is called on the render thread (right now we are doing XSync here, but with XFlush behavior will be different)
4) XPutImage request is queued to render thread's X11 connection socket, render thread does NOT wait for it to reach the X Server since XFlush is now being used
5) UI thread is notified about the last pending render operation being completed now
6) UI thread calls XDestroyWindow, request is written to the UI thread's X11 socket

Now we have a potential race condition here. If X Server processes XDestroyWindow request first, XPutImage command will arrive with now invalid XID and BadMatch error will be triggered.

**Reply by lindexi** [reply]
@kekekeks Thank you for your patience.

**Reply by lindexi** [reply]
@kekekeks  I have written a test program to illustrate the confusion I encountered. I attempted to use a Boolean variable to toggle between the calls of XFlush and XSync methods, aiming to test and compare their performance. To minimize the impact, the code I wrote is a simplified and decomposed version of the actual implementation logic in Avalonia. The issue I encountered is also the one I faced while using Avalonia. I ran the code on my device, which operates on the UOS system, and the processor model is ZHAOXIN KaiXian KX-U6780A. This processor is a low-performance one. The log output from the run is as follows, and you can see the XSync may take a lot of time:

XSync:

```
Create SKBitmap Cost: 1 ms
Copy Pixel Cost: 2  ms
XLockDisplay Cost: 0 ms
CreateImage Cost: 0 ms
XCreateGC Cost: 0  ms
XPutImage Cost: 7 ms
XFreeGC Cost: 0 ms
XSync Cost: 5245 ms
XUnlockDisplay Cost: 0 ms
```

XFlush:

```
Create SKBitmap Cost: 2 ms
Copy Pixel Cost: 2  ms
XLockDisplay Cost: 0 ms
CreateImage Cost: 0 ms
XCreateGC Cost: 0  ms
XPutImage Cost: 11 ms
XFreeGC Cost: 0 ms
XFlush Cost: 0 ms
XUnlockDisplay Cost: 0 ms
```

My test code is available at [GitHub](https://github.com/lindexi/lindexi_gd/blob/d553a7c1ac7ecce965300526a57261631b1c9ae4/X11/JaibubarebafoNilairdihaw/Program.cs). You can pull my code and run it. During the running process, pressing the Enter key tautologically will simulate triggering exposure and thus execute image rendering push.

The command line code to pull my test demo code:

```
git init
git remote add origin https://github.com/lindexi/lindexi_gd.git
git pull origin d553a7c1ac7ecce965300526a57261631b1c9ae4
```

The main code is:

```csharp
        var stopwatch = Stopwatch.StartNew();
        XLockDisplay(display);

        stopwatch.Stop();
        list.Add($"XLockDisplay Cost: {stopwatch.ElapsedMilliseconds} ms");
        stopwatch.Restart();

        try
        {
            var image = CreateImage(renderBitmap);

            stopwatch.Stop();
            list.Add($"CreateImage Cost: {stopwatch.ElapsedMilliseconds} ms");
            stopwatch.Restart();

            var gc = XCreateGC(display, handle, 0, IntPtr.Zero);

            stopwatch.Stop();
            list.Add($"XCreateGC Cost: {stopwatch.ElapsedMilliseconds}  ms");
            stopwatch.Restart();

            XPutImage(display, handle, gc, ref image, 0, 0, 0, 0, (uint) renderBitmap.Width,
                (uint) renderBitmap.Height);

            stopwatch.Stop();
            list.Add($"XPutImage Cost: {stopwatch.ElapsedMilliseconds} ms");
            stopwatch.Restart();

            XFreeGC(display, gc);

            stopwatch.Stop();
            list.Add($"XFreeGC Cost: {stopwatch.ElapsedMilliseconds} ms");
            stopwatch.Restart();

            if (useXFlush)
            {
                XFlush(display);
                stopwatch.Stop();
                list.Add($"XFlush Cost: {stopwatch.ElapsedMilliseconds} ms");
                stopwatch.Restart();
            }
            else
            {
                XSync(display, true);
                stopwatch.Stop();
                list.Add($"XSync Cost: {stopwatch.ElapsedMilliseconds} ms");
                stopwatch.Restart();
            }

            useXFlush = !useXFlush;
        }
        finally
        {
            XUnlockDisplay(display);

            stopwatch.Stop();
            list.Add($"XUnlockDisplay Cost: {stopwatch.ElapsedMilliseconds} ms");
            stopwatch.Restart();
        }
        stopwatch.Stop();
```

**Reply by kekekeks** [reply]
I think in the long run we need to use MIT-SHM extension since it provides a completion event for XShmPutImage and uses a shared memory area instead of writing the bitmap to a socket.
That way we can have backpressure from the server (i. e. we shouldn't attempt rendering a frame if X Server have 2 pending XShmPutImage requests in-flight).

**Reply by lindexi** [reply]
@kekekeks Yeah, and I wrote the demo that use XShmPutImage to replace the XPutImage. You can find my demo code in https://github.com/lindexi/lindexi_gd/blob/d87600b2a80ff3197be2a0eed70faa003cee27a1/X11/WercawchallwarnefeWhedurcachay/Program.cs

And as I test, the XShmPutImage is very fast than XPutImage in my device, my device is UOS system with ZHAOXIN KaiXian KX-U6780A CPU.

The XShmPutImage take less than 1ms but the XPutImage may take 10ms

**Reply by kekekeks** [reply]
The SHM thingy should have an API like this:
```cs
class DeferredDisplayEvents(IntPtr display)
{
     public event Action<XEvent> X11Event;
     public void ProcessPendingEvents(); // while(XPending()) {XNextEvent loop}, called periodically from code parts that rely on such events
     public void WaitForEvent(CancellationToken token); // while(!token.IsCancellationRequested) {XNextEvent loop}, this is needed to provide backpressure when we need to wait for one of pending XPutShmImage requests to complete
}

internal class X11ShmImageManager(IntPtr display, DeferredDisplayEvents events)
{
     X11ShmImage CreateImage(PixelSize size);
}

internal class X11ShmImage : IDisposable
{
     bool IsReady {get;} // Returns false if we haven't got a completion event since the last Present, can call ProcessPendingEvents here
     void WaitForReadySync(); // Waits for the last present to complete, can call WaitForEvent here
     void PresentAsync();
}

internal class X11ShmImageSwapchain(IntPtr xid, int depth, int swapchainSize = 2) : IFramebufferRenderTarget
{
    Queue<X11ShmImage> _availableQueue = new();
    Queue<X11ShmImage> _presentationQueue = new();
    PixelSize? _lastSize
    

    void DrainPresentationQueue()
    {
         while(_presentationQueue.Count > 0 && _presentationQueue.Peek().IsReady)
              _availableQueue.Enqueue(_presentationQueue.Dequeue());
    }
    
    public ILockedFramebuffer Lock()
    {
          1) gets the current window geometry, if it it doesn't match the _lastSize disposes images and clears queues
          2) calls DrainPresentationQueue();
          3) if there is an image in _availableQueue - use this image
          4) else if there are no images in _availableQueue and _presentationQueue.Count < swapchainSize - create new image
          5) else synchronously wait for the first image from _presentationQueue and use it
          6) return a framebuffer associated with chosen image
    }


}
```

**Reply by kekekeks** [reply]
Note that X11ShmImageSwapchain has to wait for all pending images to get presented on Dispose to avoid BadMatch errors

**Reply by lindexi** [reply]
I update the demo code which can re-render when receive XShmCompletionEvent event. And I test my demo and find it can render very fast and draw the touch trajectory timely. You can find my demo in [github](https://github.com/lindexi/lindexi_gd/tree/7d461891993bb3b25faa5ed4c42a9cf74edfe41d/X11/BecilanegukemCawhafebe)

**Reply by kekekeks** [reply]
Please, submit a PR to Avalonia.X11 backend. Make sure that it's a separate platform surface that can be optionally enabled via X11PlatformSettings.

**Reply by lindexi** [reply]
> Please, submit a PR to Avalonia.X11 backend. Make sure that it's a separate platform surface that can be optionally enabled via X11PlatformSettings.

Open PR in https://github.com/AvaloniaUI/Avalonia/pull/17118

I am working in my spare time
